packages/contracts-bedrock/contracts/L1/OptimismPortal.sol

```diff
@@ -185,6 +201,66 @@ contract OptimismPortal is Initializable, ResourceMetering, Semver {
             "OptimismPortal: invalid withdrawal inclusion proof"
         );
 
+        // Designate the withdrawalHash as proven by storing the `outputRoot`, `timestamp`,
+        // and `l2BlockNumber` in the `provenWithdrawals` mapping. A certain withdrawal
+        // can be proved multiple times and thus overwrite a previously stored `ProvenWithdrawal`,
+        // but this is safe due to the replay check in `finalizeWithdrawalTransaction`.
+        provenWithdrawals[withdrawalHash] = ProvenWithdrawal({
+            outputRoot: outputRoot,
+            timestamp: uint128(block.timestamp),
+            l2BlockNumber: uint128(_l2BlockNumber)
+        });
+
+        // Emit a `WithdrawalProven` event.
+        emit WithdrawalProven(withdrawalHash, _tx.sender, _tx.target);
+    }
+
+    /**
+     * @notice Finalizes a withdrawal transaction.
+     *
+     * @param _tx Withdrawal transaction to finalize.
+     */
+    function finalizeWithdrawalTransaction(Types.WithdrawalTransaction memory _tx) external {
+        // Prevent nested withdrawals within withdrawals.
+        require(
+            l2Sender == DEFAULT_L2_SENDER,
+            "OptimismPortal: can only trigger one withdrawal per transaction"
+        );
+
+        // All withdrawals have a unique hash, we'll use this as the identifier for the withdrawal
+        // and to prevent replay attacks.
+        bytes32 withdrawalHash = Hashing.hashWithdrawal(_tx);
+
+        // Grab the proven withdrawal from the `provenWithdrawals` map.
+        ProvenWithdrawal memory provenWithdrawal = provenWithdrawals[withdrawalHash];
+
+        // Ensure that the withdrawal has been proven
+        require(provenWithdrawal.timestamp != 0, "OptimismPortal: withdrawal has not been proven");
```

Comment:
Not entirely necessary but checking that `provenWithdrawal.timestamp >= L2_ORACLE.STARTING_TIMESTAMP()` might be a good sanity check to prevent weird edge cases. I don't know if it would really help (it's also more expensive) but it would've stopped something like the Nomad hack.

Reply:
RFC @smartcontracts 

https://github.com/ethereum-optimism/optimism/blob/e98e579957179b1b50a7fb3bee86b007a39480c4/packages/contracts-bedrock/contracts/L1/OptimismPortal.sol#L240-L245

+

https://github.com/ethereum-optimism/optimism/blob/e98e579957179b1b50a7fb3bee86b007a39480c4/packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol#L444-L477